### PR TITLE
frontend: data-pipelines build and test running CI/CD

### PR DESCRIPTION
**Pull request by ivakoleva**
We need to build and run unit tests of data-pipelines library and ui application that uses it. The component is dependent on shared-components library.

Added Bash scripts to
(1) link if any shared-components rebuild was made and is about to be published, or otherwise default to latest already published one 
(2) build the data-pipelines library and link @vdk/data-pipelines 
(3) build the UI application
(4) lint the subprojects to verify format
(5) run the unit tests
One bash script is meant to be used to satisfy build prerequisites, the other one to rebuild and relink continuously, needed during development.
The proposed implementation triggers a data-pipelines rebuild (+test), upon any shared-components change, as well as data-pipelines changes. 
A README.md link fix to the VEP.

Testing done: run the install script (that evaluates the build script) on a clean setup, and verified the unit tests were successful. Also, did start and browse the application manually, to check its operational.

### CODE REVIEW

File: projects/frontend/cicd/install_data_pipelines.sh
Comment (ivakoleva) [inline projects/frontend/cicd/install_data_pipelines.sh:30]: This PR depends on having the shared-components build prerequisite, upon shared-components change.
See in `gitlab-ci.yml`:
```frontend_shared_components_changes: &frontend_shared_components_locations```
https://gitlab.com/vmware-analytics/versatile-data-kit/-/jobs/3829441527#L52

File: projects/frontend/cicd/build_data_pipelines.sh
Comment (DeltaMichael) [inline projects/frontend/cicd/build_data_pipelines.sh:12]: Is this check really necessary? We already know what's in the runner image.

File: projects/frontend/cicd/build_data_pipelines.sh
Comment (DeltaMichael) [inline projects/frontend/cicd/build_data_pipelines.sh:20]: Since we're treating `data-pipelines` and 'shared-components` as separate packages, maybe we should split the CI into the subfolders instead.

File: projects/frontend/data-pipelines/gui/projects/data-pipelines/src/lib/components/data-job/pages/details/data-job-details-page.component.spec.ts
Comment (DeltaMichael) [inline projects/frontend/data-pipelines/gui/projects/data-pipelines/src/lib/components/data-job/pages/details/data-job-details-page.component.spec.ts:14]: Nit: Renaming this stuff belongs in a different PR. If we have to revert this one for some reason, we'll also be reverting the renaming. Normally I'd be against it, but it doesn't seem likely we'll revert this CI change, so up to you if you want to leave it in.

File: projects/frontend/cicd/build_data_pipelines.sh
Comment (DeltaMichael) [inline projects/frontend/cicd/build_data_pipelines.sh:21]: What's the point of executing this and the previous line?

File: projects/frontend/cicd/install_data_pipelines.sh
Comment (DeltaMichael) [inline projects/frontend/cicd/install_data_pipelines.sh:42]: Wouldn't it be better if we ran the linter as a pre-commit hook?

**Comment by tozka** [comment]
I was once told (when our open source team was reviewing our commit history) that if we have bullet points in the commit message, a change is doing too many things and should have been split. Since then when I catch myself that I make bullet points or start saying "fixing this and that" I consider if I should better split it into 2 or more. 

I thought that was good advice and I am passing it on.

File: projects/frontend/cicd/install_data_pipelines.sh
Comment (tozka) [reply projects/frontend/cicd/install_data_pipelines.sh:42]: Yes I think we should. We should also use auto formatted as in python and java

File: projects/frontend/cicd/build_data_pipelines.sh
Comment (ivakoleva) [reply projects/frontend/cicd/build_data_pipelines.sh:12]: The script is meant to be reused during local development as well, see #L6
```
###
 # For installing the build prerequisites, consider `install_data_pipelines.sh` script.
 # This script is meant to be used, to quickly
 # rebuild and link the the @vdk/data-pipelines in an already bootstrapped development setup.
 ###
```
In both cases it's good to have a message asserting on prerequisites known, to reduce on debugging complexity

File: projects/frontend/cicd/build_data_pipelines.sh
Comment (ivakoleva) [reply projects/frontend/cicd/build_data_pipelines.sh:20]: Those npm artifacts separated, are something to push faster and unblock adoption. We will next be working on having a potentially image, that contains the frontend installation, so the Frontend component (the way described VEP-1507), is going to be delivered at once to users
And the compounds of the Frontend could be also consumed;

So, in that sense, and also considering the reuse of commons in gitlab-ci.yml between subprojects, for example change locations to trigger pipeline jobs, I still think it's good to continue having common `cicd/` dir for the Frontend component.

File: projects/frontend/data-pipelines/gui/projects/data-pipelines/src/lib/components/data-job/pages/details/data-job-details-page.component.spec.ts
Comment (ivakoleva) [reply projects/frontend/data-pipelines/gui/projects/data-pipelines/src/lib/components/data-job/pages/details/data-job-details-page.component.spec.ts:14]: Could be, however the build is actually introduced in this pr, so I would rather have the build to compile the code changed.

Comment (ivakoleva) [reply projects/frontend/data-pipelines/gui/projects/data-pipelines/src/lib/components/data-job/pages/details/data-job-details-page.component.spec.ts:14]: I considered in case of separation, should the code import fixes be before or after the build script.
In case before, I still cannot verify if the fixes are operational. In case after, the build cannot be merged, due the code fails with compile errors. So I decided to have them both simultaneously.

File: projects/frontend/cicd/build_data_pipelines.sh
Comment (ivakoleva) [reply projects/frontend/cicd/build_data_pipelines.sh:21]: L20 so that the script navigates on FS correctly, independently of where it is executed from
L21 is to navigate, so that other steps are relative to this one (data-pipelines component source root)

**Comment by ivakoleva** [comment]
> I was once told (when our open source team was reviewing our commit history) that if we have bullet points in the commit message, a change is doing too many things and should have been split. Since then when I catch myself that I make bullet points or start saying "fixing this and that" I consider if I should better split it into 2 or more.
> 
> I thought that was good advice and I am passing it on.

I generally agree, it's a good advice. When you are open sourcing an existing component, a lot of stuff are known and also tested & verified already in advance. So, I also think it is normal to be confident with enumerating the changes, it rather depends on the context.

**Comment by tozka** [comment]
> 
> I generally agree, it's a good advice. When you are open sourcing an existing component, a lot of stuff are known and also tested & verified already in advance. So, I also think it is normal to be confident with enumerating the changes, it rather depends on the context.

Yes I agree, for that specific case, we are fine.

 I wanted to note this advice because the change reminded me. 
And also want to note (especially for any new contributors) that while now we have some bigger PRs around open-sourcing the UI and that's not a practice we recommend.

**Comment by ivakoleva** [comment]
> > I generally agree, it's a good advice. When you are open sourcing an existing component, a lot of stuff are known and also tested & verified already in advance. So, I also think it is normal to be confident with enumerating the changes, it rather depends on the context.
> 
> Yes I agree, for that specific case, we are fine.
> 
> I wanted to note this advice because the change reminded me. And also want to note (especially for any new contributors) that while now we have some bigger PRs around open-sourcing the UI and that's not a practice we recommend.

I actually think 11 files is a pretty standard amount of file changes. The PR description does not contain bullets, it contains logical steps that the install script performs, that would be an atomic change (since the script is an atomic item itself). So I don't think artificially splitting the script, in a series of PRs, would be practical. I considered splitting the import fixes, but decided not to due arguments listed in https://github.com/vmware/versatile-data-kit/pull/1671#discussion_r1118690256

File: projects/frontend/cicd/build_data_pipelines.sh
Comment (ivakoleva) [reply projects/frontend/cicd/build_data_pipelines.sh:21]: It's a standard practice, we use it a lot in VDK codebase

Comment (DeltaMichael) [reply projects/frontend/cicd/build_data_pipelines.sh:21]: Makes sense when you take into account local reusability.

**Review by DeltaMichael — APPROVED** [review]

File: projects/frontend/cicd/install_data_pipelines.sh
Comment (ivakoleva) [reply projects/frontend/cicd/install_data_pipelines.sh:42]: thanks for pointing this, I am working to configure a pre-commit hook 😊

Comment (ivakoleva) [reply projects/frontend/cicd/install_data_pipelines.sh:42]: I've opened a new tech stack code formatting PR https://github.com/vmware/versatile-data-kit/pull/1684, that I want to merge after this (so we have build and testing available)

Comment (ivakoleva) [reply projects/frontend/cicd/install_data_pipelines.sh:42]: Do you have a pre-commit hook in mind, 
for running the npm lint? (asking since lost more than 15 minutes to find such ootb)

**Review by tozka — APPROVED** [review]

File: projects/frontend/cicd/.gitlab-ci.yml
Comment (ivakoleva) [inline projects/frontend/cicd/.gitlab-ci.yml:24]: The job needs `frontend-shared-components-build` for:
1. exporting `artifact: ..dist/..`
2. running also in case of `$CI_COMMIT_BRANCH == "main"`
we need to add on it 
cc @DeltaMichael

Comment (DeltaMichael) [reply projects/frontend/cicd/.gitlab-ci.yml:24]: Not sure what you mean by this. Maybe you can call the build_shared script as part of this job if that's what you're referring to.

Comment (ivakoleva) [reply projects/frontend/cicd/.gitlab-ci.yml:24]: On release, we need to publish the same artifact build and already tested on build stage. Check out my changes, I'm resolving this  feel free to open if needed

File: projects/frontend/cicd/.gitlab-ci.yml
Comment (DeltaMichael) [inline projects/frontend/cicd/.gitlab-ci.yml:90]: Where does the package we publish to npm registry come from if we don't run the build script?

File: projects/frontend/cicd/.gitlab-ci.yml
Comment (DeltaMichael) [inline projects/frontend/cicd/.gitlab-ci.yml:62]: Where does the package we publish to npm registry come from if we don't run the build script?

Comment (DeltaMichael) [reply projects/frontend/cicd/.gitlab-ci.yml:62]: Ah, nvm, just saw that you run the build stage on merge to main as well.

Comment (ivakoleva) [reply projects/frontend/cicd/.gitlab-ci.yml:62]: The package is built once during build stage, then the artifact is reused by other jobs (until expired in 7 days).

**Pull request merged**
